stan/math/rev/core/var.hpp

```diff
@@ -313,6 +316,16 @@ class var_value {
 // For backwards compatability the default value is double
 using var = var_value<double>;
 
+template <typename MatrixType>
+class arena_matrix;
+
+template <typename MatrixType>
+class arena_matrix<var_value<MatrixType>> : public var_value<MatrixType> {
+ public:
+  template <typename T>
+  arena_matrix(T&& x) : var_value<MatrixType>(std::forward<T>(x)) {}  // NOLINT
+};
```

Comment:
So without this constructor I'm getting a conversion error since `arena_matrix` doesn't understand how to convert between a `var<matrix>` and `arena_matrix<var<matrix>>`.

```cpp
./stan/math/rev/fun/softmax.hpp:33:27: error: conversion from ‘const stan::math::var_value<Eigen::Matrix<double, -1, 1> >’ to non-scalar type ‘stan::math::arena_matrix<stan::math::var_value<Eigen::Matrix<double, -1, 1> > >’ requested
   arena_matrix<mat_plain> alpha_arena = alpha;
                           ^~~~~~~~~~~
./stan/math/rev/fun/softmax.hpp:35:27: error: conversion from ‘arena_matrix<Eigen::Matrix<double, -1, 1>>’ to non-scalar type ‘arena_matrix<stan::math::var_value<Eigen::Matrix<double, -1, 1> >>’ requested
   arena_matrix<mat_plain> res = res_val;
```

Since this adds the implicit `operator=` for converting a `var<matrix>` to an `arena_matrix<var<matrix>>` and everything is actually handled by `var_value<Matrix>`. `var<matrix>` also has constructors to take in `arena_matrix<var<matrix>>` so we don't copy mem

Reply:
You should only use arena matrix for Eigen types. `var<matrix>` is not an Eigen type so you should use `arena_t` metaprogram to determine appropriate type for arena storage (which turns out to be `var<matrix>` itself).

Reply:
Oh that's a much nicer way to do it!